pkg/webhooks/admission/queues/mutate/mutate_hierarchical_queue.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2024 The Volcano Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutate
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	whv1 "k8s.io/api/admissionregistration/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/klog/v2"
+
+	busv1alpha1 "volcano.sh/apis/pkg/apis/bus/v1alpha1"
+	vcbus "volcano.sh/apis/pkg/apis/bus/v1alpha1"
+	"volcano.sh/apis/pkg/apis/helpers"
+	schedulingv1beta1 "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+	"volcano.sh/volcano/pkg/webhooks/router"
+	"volcano.sh/volcano/pkg/webhooks/schema"
+	"volcano.sh/volcano/pkg/webhooks/util"
+)
+
+func init() {
+	router.RegisterAdmission(hierarchyService)
+}
+
+var hierarchyService = &router.AdmissionService{
+	Path:   "/hierarchicalqueues/mutate",
+	Func:   HierarchicalQueues,
+	Config: config,
+
+	MutatingConfig: &whv1.MutatingWebhookConfiguration{
+		Webhooks: []whv1.MutatingWebhook{{
+			Name: "mutatehierarchicalqueues.volcano.sh",
+			Rules: []whv1.RuleWithOperations{
+				{
+					Operations: []whv1.OperationType{whv1.Create, whv1.Update},
+					Rule: whv1.Rule{
+						APIGroups:   []string{schedulingv1beta1.SchemeGroupVersion.Group},
+						APIVersions: []string{schedulingv1beta1.SchemeGroupVersion.Version},
+						Resources:   []string{"queues"},
+					},
+				},
+			},
+		}},
+	},
+}
+
+var config = &router.AdmissionServiceConfig{}
+
+// HierarchicalQueues mutate queues when capacity plugin enables hierarchy.
+func HierarchicalQueues(ar admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
+	klog.V(3).Infof("Mutating hierarchical queue %s for operation %s.", ar.Request.Name, ar.Request.Operation)
+
+	queue, err := schema.DecodeQueue(ar.Request.Object, ar.Request.Resource)
+	if err != nil {
+		return util.ToAdmissionResponse(err)
+	}
+
+	switch ar.Request.Operation {
+	case admissionv1.Create, admissionv1.Update:
+		err = closeQueue(queue)
+	default:
+		return util.ToAdmissionResponse(fmt.Errorf("invalid operation `%s`, "+
+			"expect operation to be `CREATE` or `UPDATE`", ar.Request.Operation))
+	}
+
+	if err != nil {
+		return &admissionv1.AdmissionResponse{
+			Allowed: false,
+			Result:  &metav1.Status{Message: err.Error()},
+		}
+	}
+
+	return &admissionv1.AdmissionResponse{
+		Allowed: true,
+	}
+}
+
+func closeQueue(queue *schedulingv1beta1.Queue) error {
```

Comment:
Maybe put these logic to volcano controller is better.

Reply:
I also think that the logic for open/close queue can be added to the controller, but do you think it's necessary to include this logic in the webhook as well for fallback handling